pkg/sql/catalog/tabledesc/logical_replication_helpers.go

```diff
@@ -0,0 +1,169 @@
+							pgcode.InvalidTableDefinition,
+							"table %s has a virtual computed column %s that is a key of index %s",
+							dst.Name, col.Name, idx.Name,
+						), cannotLDRMsg)
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// Decoding a primary key with a composite type requires reading the current
+// value. When the rangefeed sends over a delete, however, we do not see the
+// current value. While we could rely on the prev value sent over the rangefeed,
+// we currently have no way to handle phantom deletes (i.e. a delete on a key
+// with no previous value)
+func checkCompositeTypesInPrimaryKey(dst *descpb.TableDescriptor) error {
+	for _, compositeColID := range dst.PrimaryIndex.CompositeColumnIDs {
+		for _, keyColID := range dst.PrimaryIndex.KeyColumnIDs {
+			if compositeColID == keyColID {
+				colName := ""
+				for _, col := range dst.Columns {
+					if col.ID == keyColID {
+						colName = col.Name
+					}
+				}
+				return errors.WithMessage(pgerror.Newf(
+					pgcode.InvalidTableDefinition,
+					"table %s has a primary key column (%s) with composite encoding",
+					dst.Name, colName,
+				), cannotLDRMsg)
+			}
+		}
+	}
+	return nil
+}
+
+// Replication does not work if a column in a family has a not null
+// constraint. Even if all columns are nullable, it is very hard to
+// differentiate between a delete and an update which nils out values in a
+// given column family.
+func checkColumnFamilies(dst *descpb.TableDescriptor) error {
+	if len(dst.Families) > 1 {
+		return errors.WithMessage(pgerror.Newf(
+			pgcode.InvalidTableDefinition,
+			"table %s has more than one column family", dst.Name,
+		), cannotLDRMsg)
+	}
+	return nil
+}
+
+// All column names and types must match with the source table’s columns. The KV
+// and SQL write path ingestion side logic assumes that src and dst columns
+// match. If they don’t, the LDR job will DLQ these rows and move on.
+func checkSrcDstColsMatch(src *descpb.TableDescriptor, dst *descpb.TableDescriptor) error {
+	if len(src.Columns) != len(dst.Columns) {
+		return errors.WithMessage(pgerror.Newf(
+			pgcode.InvalidTableDefinition,
+			"destination table %s has %d columns, but the source table %s has %d columns",
+			dst.Name, len(dst.Columns), src.Name, len(src.Columns),
+		), cannotLDRMsg)
+	}
+	for i := range src.Columns {
+		srcCol := src.Columns[i]
+		dstCol := dst.Columns[i]
+
+		if srcCol.Name != dstCol.Name {
+			return errors.WithMessage(pgerror.Newf(
+				pgcode.InvalidTableDefinition,
+				"destination table %s column %s at position %d does not match source table %s column %s",
+				dst.Name, dstCol.Name, i, src.Name, srcCol.Name,
+			), cannotLDRMsg)
+		}
+
+		if dstCol.Type.UserDefined() {
+			return errors.WithMessage(pgerror.Newf(
+				pgcode.InvalidTableDefinition,
+				"destination table %s column %s has user-defined type %s",
+				dst.Name, dstCol.Name, dstCol.Type.SQLStringForError(),
+			), cannotLDRMsg)
+		}
+
+		if !srcCol.Type.Identical(dstCol.Type) {
```

Comment:
so you're saying that `Idential` returns false for 2 User defined types, even if they have same definition on source and destination?

Reply:
if we want to handle user-defined types for LDR, then in the if statement right above this one, we could make a separate check that the UDT definitions are equivalent, so we avoid using `Identical` for them.

Reply:
ok carry on.

Reply:
ah, stale read! If you think this check would be easy to create and accurate, it would be great to add it!

Reply:
scratch that; i was too optimistic. doing what i said above would require also loading the type descriptor from the source cluster, which is something we should punt on.

Reply:
ok, no worries!